<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Jayaram Kancherla, Héctor Corrada Bravo" />

<meta name="date" content="2016-06-15" />

<title>Introduction to epivizr: interactive visualization for genomic data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Introduction to <code>epivizr</code>: interactive visualization for genomic data</h1>
<h4 class="author"><em>Jayaram Kancherla, Héctor Corrada Bravo</em></h4>
<h4 class="date"><em>2016-06-15</em></h4>



<p><a href="http://epiviz.cbcb.umd.edu"><code>Epiviz</code></a> is an interactive visualization tool for functional genomics data. It supports genome navigation like other genome browsers, but allows multiple visualizations of data within genomic regions using scatterplots, heatmaps and other user-supplied visualizations. It also includes data from the <a href="http://barcode.luhs.org/">Gene Expression Barcode project</a> for transcriptome visualization. It has a flexible plugin framework so users can add <a href="http://d3js.org/">d3</a> visualizations. You can find more information about Epiviz at <a href="http://epiviz.cbcb.umd.edu/help" class="uri">http://epiviz.cbcb.umd.edu/help</a> and see a video tour <a href="http://youtu.be/099c4wUxozA">here</a>.</p>
<iframe width="480" height="360" src="http://www.youtube.com/embed/099c4wUxozA" frameborder="1" allowfullscreen>
</iframe>
<p>The <code>epivizr</code> package implements two-way communication between the <code>R/Bioconductor</code> computational genomics environment and <code>Epiviz</code>. Objects in an <code>R</code> session can be displayed as tracks or plots on Epiviz. Epivizr uses Websockets for communication between the browser Javascript client and the R environment, the same technology underlying the popular <a href="http://www.rstudio.com/shiny/">Shiny</a> system for authoring interactive web-based reports in R.</p>
<div id="preliminaries-the-environment" class="section level3">
<h3>Preliminaries: the environment</h3>
<p>The R programming language is becoming the tool of choice for exploratory and data analysis in genomics. A major reason for that is the large <a href="http://bioconductor.org/">Bioconductor</a> project, which provides software infrastructure for many common genomics data analysis tasks, and hosts a large number of third-party contributed data analysis tools and methods. For modern large-scale genomics data analysis, a large number of the state-of-the-art methods are implemented as R packages hosted by the Bioconductor project. The Bioconductor project also provides a lot of materials for learning how to use packages and infrastructure: <a href="http://bioconductor.org/help/course-materials/">Bioconductor courses and conferences</a> The Bioconductor community is vibrant and active, including both users and software developers. <a href="https://support.bioconductor.org/">Bioconductor support site</a></p>
<p>R is freely available as a binary for most operating systems at <a href="https://cran.r-project.org/index.html" class="uri">https://cran.r-project.org/index.html</a>. Follow instructions there to download and install R.</p>
<p>A commonly used development environment is Rstudio, also freely available for most operating systems at <a href="https://www.rstudio.com/products/RStudio/#Desktop" class="uri">https://www.rstudio.com/products/RStudio/#Desktop</a></p>
<p>To begin using Bioconductor, start up Rstudio and type in the console pane:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&quot;http://bioconductor.org/biocLite.R&quot;</span>)</code></pre></div>
</div>
<div id="preliminaries-the-data" class="section level3">
<h3>Preliminaries: the data</h3>
<p>In this vignette we will look at colon cancer methylation data from the TCGA project and expression data from the gene expression barcode project. The <code>epivizrData</code> package contains human chromosome 11 methylation data from the Illumina 450kHumanMethylation beadarray processed with the <code>minfi</code> package. We use expression data from the <code>antiProfilesData</code> bioconductor package.</p>
<p>To install these packages we use the <code>biocLite</code> package install command provided by Bioconductor</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biocLite</span>(<span class="kw">c</span>(<span class="st">&quot;epivizr&quot;</span>, <span class="st">&quot;antiProfilesData&quot;</span>, <span class="st">&quot;SummarizedExperiment&quot;</span>))</code></pre></div>
<p>This will install these packages, and other packages they depend on.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(epivizr)
<span class="kw">library</span>(antiProfilesData)
<span class="kw">library</span>(SummarizedExperiment)</code></pre></div>
<p>To get a sense of what these data are you can explore an epiviz workspace for a different dataset at <a href="http://epiviz.cbcb.umd.edu/?ws=Gx3GEFT4iti" class="uri">http://epiviz.cbcb.umd.edu/?ws=Gx3GEFT4iti</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(tcga_colon_blocks)
<span class="kw">data</span>(tcga_colon_curves)
<span class="kw">data</span>(apColonData)</code></pre></div>
<p>The <code>tcga_colon_blocks</code> object is a <code>GRanges</code> object containing chromosome 11 regions of hypo or hyper methylation in colon cancer identified using the <code>blockFinder</code> function in the <code>minfi</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">show</span>(tcga_colon_blocks)</code></pre></div>
<pre><code>## GRanges object with 129 ranges and 11 metadata columns:
##         seqnames                 ranges strand |               value
##            &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; |           &lt;numeric&gt;
##     [1]    chr11 [  4407026,   6435089]      * |  -0.142954896408814
##     [2]    chr11 [131239366, 133716186]      * |  -0.135137323912239
##     [3]    chr11 [ 55041873,  57022542]      * |   -0.17334711677526
##     [4]    chr11 [114645223, 116602403]      * |  -0.140934365473709
##     [5]    chr11 [ 78357700,  80184550]      * |   -0.15603691753532
##     ...      ...                    ...    ... .                 ...
##   [125]    chr11   [29644815, 29650449]      * | -0.0940450729648221
##   [126]    chr11   [41943963, 41956273]      * | -0.0760193132782017
##   [127]    chr11   [16298618, 16314417]      * | -0.0748443876820716
##   [128]    chr11   [38740154, 38740154]      * |  -0.117248761155248
##   [129]    chr11   [81757203, 81757203]      * | -0.0710557720226853
##                       area   cluster indexStart  indexEnd         L
##                  &lt;numeric&gt; &lt;numeric&gt;  &lt;integer&gt; &lt;integer&gt; &lt;numeric&gt;
##     [1]   30.3064380386685       495     130755    131000       212
##     [2]   23.9193063324663       520     141959    142173       177
##     [3]   19.5882241956043       507     134251    134374       113
##     [4]   14.0934365473709       520     140000    140138       100
##     [5]   14.0433225781788       507     138132    138249        90
##     ...                ...       ...        ...       ...       ...
##   [125]  0.188090145929644       497     132733    132734         2
##   [126]  0.152038626556403       502     133379    133380         2
##   [127]  0.149688775364143       495     131986    131987         2
##   [128]  0.117248761155248       499     133331    133331         1
##   [129] 0.0710557720226853       508     138263    138263         1
##          clusterL            p.value      fwer       p.valueArea  fwerArea
##         &lt;integer&gt;          &lt;numeric&gt; &lt;numeric&gt;         &lt;numeric&gt; &lt;numeric&gt;
##     [1]      1629                  0         0                 0         0
##     [2]      1759                  0         0                 0         0
##     [3]      1816                  0         0                 0         0
##     [4]      1759                  0         0                 0         0
##     [5]      1816                  0         0                 0         0
##     ...       ...                ...       ...               ...       ...
##   [125]       367 0.0949029198604193         1 0.416952488890214         1
##   [126]        45  0.316487220018491         1 0.485758597035402         1
##   [127]      1629  0.342091920427093         1 0.494228876494974         1
##   [128]         7 0.0564138506964121         1 0.592024814339825         1
##   [129]        22  0.729905454979272         1 0.860076351814847         1
##   -------
##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The columns <code>value</code> and <code>p.value</code> can be used to determine which of these regions, or blocks, are interesting by looking at a volcano plot for instance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(tcga_colon_blocks$value, -<span class="kw">log10</span>(tcga_colon_blocks$p.value), <span class="dt">main=</span><span class="st">&quot;Volcano plot&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Avg. methylation difference&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;-log10 p-value&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(-.<span class="dv">5</span>,.<span class="dv">5</span>))</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAYAAAAGACAMAAACTGUWNAAADAFBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////isF19AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO2dd0AT1x/AXwKEEVYYYQkIiqKCqCDOioijKmJxVnH8FLUibtx7L9yjahUtKuLEUbRV21q1rqpVW1tbd90TBwooJO93dwmQICG5+WJ4nz9yI++97+U+cHfv7t57AGKQAlBvQHkHC0AMFoAYLAAxWABisADEYAGIwQIQgwUgBgtADBaAGCwAMVgAYrAAxGABiMECEIMFIAYLQAwWgBgsADFYAGKwAMRgAYjBAhCDBSAGC0AMFoAYLAAxWABisADEYAGIMWIB6wCYRs30A+BHzS9aAUvWhXNRBicYsYDHItCQnCp9gNMHzS94EXBu6tTrrEtlgBELgI2B2Uti8i8A/9Naz4uA1QB8z7pUBhizgEUA7CYmKwDYr7UeCxCIWwAMICYxQJoLYe7UBvY+rQ+R61U778WgEJuqic+JOeXOxnLrgMR7xGwo8Mrq7eMSe5tMp5GFwguEXouRhUzJLyqjOEUUIKgn8A8kMWYBMARUVMIPdqALhM+rAYoJUL3zbntSyxVeQLhU9VWVN6QAp1Bq/r12Fgov4OdPrvhcoS5DI0VPFwDcOyL4jUYtYCoA/8ITAGyDMAGAyPQljkB0Ub3zuhH/HhmDAUiA+bbAd8m25gDsIwUAt6lz5AB8p52FwgsAu5kLPakDG1WGZgp8CCqFSwCsgFOA5DV8JwZViSPHEQDiVTsvi/hDhlARAPzg3bZtMyA8DMBCSsAfEG4CYJF2FgpCwAEI/wSghaoMrRRYQCko/UA0rA/aQngZgOnkCjfQWLXzzgOwlFjx4M8/lYSH86sH+gCQTApwJdaeBWCmdhYKL+BMJIa1gI+qDK0UWEBpjATSp2KQAmEmAOvIFXVABdXO2wNAamGqdDfiOC5TC/AiVpwjBWhlofACNchJG2CupMrQSoEFlAZx/B8KxM+K/wPcQX2VgNPkUUbFVTNQdfXVX0sK0MpCof4PqE0ctrT+A6gUWEBpFLgCCxBJzBCH68B86kDfR7XznhHnTyVUfu5VR7kGgB0QLiwpQCsLheocQJxY2hSdA4pTEAIOoviJxi2AvA1EnIcJBgLQbBtxwSK+oL6CiSF22/ZeAPSH3wLQZPsCaUkBWlkoCAH2s5LdAfhBXYZminUAzM5H8AuNXMABQgBZwYLPA1WX7FOgeuf960ot+72Ad23JGeL6Z6K2AM0sFF7AzZpc0UVZWA/QSHEcV8RKI88OhKvmcifXs/NuQ90VVdVin/SrYR04IouYO1bfJmTBKwvgmKMlQDMLBVETPhxhHzJHUVSGRgrlOBcsgGcIAag34WOwAMRgAYjBAhBTngQYJVgAYrAAxGABiMECEIMFIAYLQAwWgBgsADFYAGKwAMRgAYjBAhCDBSAGC0AMFoAYLAAxWABisADEYAGIwQIQgwUgBgtADBaAGAYC8q89VnK/IeUVOgKexf8OYcEMGwAq7uJtg8obdATcBnsgXAD6ZuztS7ZxwHABbQG+ieRs70aaX+SeL6RjKIai8WOeBOSDTHI2zVbzi1NFcc1GnMeQBFzmSQAMWEnOLvAqPYntOhrlmTI1+RHgGdU/wvkGVOyXDSo9CRaghhcBubuTE1oFWKTD70Hrl6Un4V3ApTHdvviyV5S7re8aniOxgxcBKgry4N1LuioCfAuY4zWlmo+dJTCv5yYKyuM3Fit4FFAmPAs45/10dG9lU7AkpT4cK56iPwMyTFTAuGnQ/yq0tE5WetwtsPDlNRY7TFRAv3XQ7g0UeYyGoeehox2vsdhhogJmjYDBZ6Cd+ao82bO3ZtV5jcUOExVw2/XsssYv24O4IZ0+NDbbwGssdpioAJjpFRNqaesCxC7monj9ydFhqgLg652Lvv56xeq2Nbsh6WbVYExWwKcCFoAYLAAxWABisADEYAGIwQIQgwUgBgtADBaAGFMXkH/nvTCBmGLaAl4NlfradLknRCimmLQARZN+L2HeDP83AsRiikkLOBRKvRMQt1SAWEwxaQFzxlGTTT0EiMUUkxYwP4mapPxPTzqUmLSAUwHUJVBr/EjyY4S5Cvqy+e/513qGf9CfEhmmLSB/eaC534S3QoRiimkLoDDq/W/yAt5NlEudR74SJhgTTFyAIrL7Xfgwvq7xngVMXMB34VRVrPkWQaIxwcQFTJwBf+oSGts3UZBoTDBxARNmTgxIPb/Du6og0ZjAq4APe57p+kogAftqepHDJzX3PiBIOAbwKuAlOKrrK6FOwv7VHsFnA0MXDhYkHAP4ERCnojNoHhdXegqhLkPHNZY5OQzJSu0lTDj68COgGbBtRFAPBDXSaqidfaQQ62/olMeIO+NjB/20tS18QcwPmst7OIbwI0DxtW2vF6Ucgn5rXojZQjrlMWGbfGLGiip9/FcSF6K73A1tjy44fJ0DbkW670N6Dnju+hfxmRP8hYdz86runRbc5zkeU3g7CStWSXvcQCggvQP5ud2+yso+1tKRq4bKjbQuxuNV0K1IGUIBK8m6132XdS3gn3L3CxD+43qb54jM4PMyVJGSdEPXd7wLOBBBfKyKX5gAJ0+cMYaYT1zMc0RmmGxNOK/yBggnDfG4APt9s5l8KJw8iueIzDAFAa/TZ6V93CXFXyH1h9S2SoNw+pjJE4jl/su5i8ghJiDge68OEzt57P9ofcGR5SkupyG85uxMXBCdc33IWUQu+fQFPHAldjKxf/8r7cvvPbvNird36DMrzv07rgJyy6cvYMkAajJ0XqnfvkqZuPbRk7UT12dxFY9jPn0Bw1VXN6sHlv618ofZyb9xFYsHGAn4cOcd2w5BuRMwdyQ1GT+11G+fRYRNGOXfw6QeSW6uLAYX+8xkp4A7Af+4kUf/+x5XSv224yhiO3PbTeMqGufQF5AGvtwHLq4Rs3vjlcOroNVu4zdNciv9IvOlQy45+duPs2hcQ19AjV7EwkU4OpBVXC7rAden9JzyT+lf/V2NmhSYG20fyvQFWG+jBGy3ZhVXoAcyzx2po//NCoJEYwJ9AbWGUgIm12QVV6gnYq3IJzGKHqOFicYA+gLWicZeAYdSbJaxiiuUgLs1Wq9IDm1pvK8n0hegXOgAALCZYixXQXrI3zR4tJFWgimY1APenNl5Ukd/rAaDm6mq+fRrwp849AW0KoRVXCxADX0B8QR929gHz2QVFwtQw/QQlFUvhVVcLEAN43NARgiruFiAGsYCNktZxRVKwOP0JT8UCBOKGfQFbKSY79qMVVyBBKyTdx3esObfgsRiBn0BUgr7puz6Q+VcgDKtlsQ7ocTr8Cd8bhKfKVWM93GA6dQDxoUee383yf+F1sreq4iPW9Ncan8x5leO43GFqQi450I+9M0Jc5J5J97/8dsTquN+xC8QrnfrKvF1aOOfYJx3pGkKWKEBq7hcC9jekfh44m/lvLiHpTisd2gt6nd1TIM35X95NNjzyOt0WDq3ETmCpgA3DVjF5VpA2pfEx1cBgxrD2w7OEcRFmk82sSK9bu6iwQfqur2GUydsi+U2IkeYyiHoH89cCCtYjR4Bp4zytX0NYUwasVbZr0pU98+t9hMn4r6nGnAbkSMYC7iUzCou5yfhnu0eQDsL97uwe1qo/B6EU1XP4X9u6tc7lJiOmpHSjeOI3EBfgGLL6CSCUE9WcTkXkDfNwddMshfCxHn2TvkQDlQ/MHrsejQgFV50PeX/I8cRuYG+gCkg2NKtrlR2iFVcHipiijsb3Gs/gHvtvCZD+MDtpnr1kQpRLnJplHwJ5wE5gb4A3wSYFg2fBnzPKi4/NeHlUomrpZlszY+LvYr3d3b6tBGT1htr14n0BVhshQ/lSri0UVmpVXzQ3V0hT7ciXu6esenRkT5RAy/yUjwP0BfgPpW4Gr0MU8u8GXege8d05RgJqLhRR/VHqJtxDzK2XRMmEkPoCxggWwfb9L3bLqCMxFtBQDOrQVbjvukOdpaegjcBtyZ1G1X0Mq5yijy2q1ffXJ5icQF9Aa/jv4CXbIDZ1jISh3QqgGmAfCWnT7jm+qwdhVitZrCxBrDJbcLYlg491UvLGj6B8G3XBH5icQLDesCrw7fKSky+PfcEHCfmtmqNqH2xcyHmi2htpqHcdzkXETZpuKSJ6o++6nnyM8vh1R/Hn/MSjz30BfQ/ptCb2IeopeXPJdsELdLxUiBPh6C1vbuOIM46CyuNJZeU5uQ9uTeTJBaOoU6J73iJyBb6AjyB97g/9SQe4rj6KjnN/9lNR6epPAmYPdqOfAluR2sPatH5EVETq9ROvGq5W3r37rxEZAuDmvCpJD8QklzmdfWbOBBBTmuAhjo6aeBJwJYW/uRk2hgL6hlMn3EQDhodWgnCM57Zrnd5CckSRucA5e8TA0WRZSZ/+S/5mX5S1+GKJwGvPWyIo/8fbj/LqcXH1bpkeHYzJ19NDL7Y+iAvIVnC7CScldZFYsYqLl+XoWek/kmd3Lf3Hq5azFsSa5lUk7wsbXyi4TF+QrKDgYDby6PMLVqnsLus4K0e8F+lyj2nhLbIphbOdK4RVTV9LHEZ+kZ2Qm6U1QH6AkKApF0q60af/NWEC7aNHJupml3pk3LlSD3LA/6f72j+mUdZFRd00BcQm/aag7g834q4Nza67+6nzieuK6CyitjKCljGGundIUbngIIk1t0f8SLgfLRXpf4PiJn98gnfrQ+tZOVf2T3lrFP4hWN3enbhIR4XMBKQD1j/OfEhIMPz2/s3pnrcgG/dzhGL31oGExdEQXVHNSYW8txv6suOBhMSoPCibsIlfwl/oN7aCxgmJS6CL1mMoNrQx+zjPCAnmJCAq6r7sw/lcCv5+PetzW92w4gLHzMZ2XkcjDDOJ5LMBCiOsB4YigcBl6h2m1eSLKYeqUHMvLdaG9vNrUWwqCXMX1rbwXypcb6jy0RA1q/bjrG9uciDgHcOj6ByuLWVjZN5hXlKCENdfoH3Dy+tVaNngyYTPCZFduI8IhcwuBc0zhoAYDVG/z3RsuDjJDzls1tbHNpWOAJnWgQF92/vZL/kv/sr5Sfedbexk3rH3w76ifuQ7KEvYCYY+UfWlVFgNqu4fAhQJDvZiP3JGz/ejU5+s+fF9S+9K3S6CmEn25W3rs1wS5zEfUj20BcQoOqLf1AVVnH5qYh98Ayjpi3dNdc6UVu8JtAom8vTF+Cs6gF1i4xVXJ5qwmGqK6EqmuMFPLJrHuMsa33SLI2XkCyhL6BLF+pFhx7s3nXl64mYhLzlucN+ssa62+5m7R9mrZWKjPKRGH0Bf1X+PP34ro62mecIGMflSYCitVnU8M+sq2drrMs3nx3rUNM+xJKXiGyhLwBowjguVwL+6eLr2+q4xopdYR7Bi7SaJL03D7/x6vKJABu23SvwAn0BtzVhHJcjAcfl8zvZ+4irXygjTY71YpfgQK80l6echOSYT7x9QJ29TYa8g1uquZXyvDe3sKZS7fj7C/MSe8mNso3Spy3glf0vwUryrsOo/gevau1f5ZZAG9uYf4m//xy403939TajXP1aGGOvQZ+2gCeuqwYRE6VVDevWleprviwzu9Zpxdvlbql1bG1rH86wtvPyTVf0TOIiJsd82gKUnjPI0Xl+sujXASo3eBU/KH3lSPUU3Vt6UKE46L1JltGz7eBLRtlzHAMBytcv2B9NOToHrA50eg4vuNeNJl817VnccPPoZ9SkMrXLj/q7+C3PXOCVbIx9J9IVcH1sZUsAJP5jdA7NYBhcXYaus5V4uDcMakAOnb16IHy699vfydVHI8jPbJvK729+gEob8RNi6aGLCfwH/GbvOWDpptRlg3wdyrry0w9nFbGCtQ29vKKoocvnjFopb987oDWxt186km/kZUv8bfxt+jyRVBtFXBHl+rXkKCaX0BTQNEJdx8xrb0SddZzxJ7cqp/qsKreJiu/Y5sTC1LAL8P0KUdt3qZXNxZIh1bzaDfCrPYbDmFxBU4D9+sLl7xxYxeX2VsTg4B1XMur0+ZxqDVLgTVx9KlP8HKRBIV6dQ0/uktiJrO3MJG06rtdXDgJoCqgbV3geGxGuO60BMBWg452kfe1qRO+ElVWtkdqqRu589SFxRaa5tbRRDS9RBHzaoKqVMd6LoClghyhq3cm/r57aEC3W0fbIQBgJyB7jZO84TPduDFd1iRJ6Wr08bPEPLT68PG9ZpackamBtiasxthajexWUGam6CxeZyS4uEwGKz3o9hE8HhGndantxaPftwvnZncj/z6M++erlXY0ORvaxDQDWsx39KvW3MMo3U+jXA55fPnz4EusGP0wE7Gvw8fjkS+UtO7j3Vd/pz4mMSDswRl707FcR2cZswOnmFk5AGhMiFYf/zGqT+eFTqgmPn0VNlmuMDbw5+A6Eb+P6qBcVm7q3Ga8xWlJuL3OR38IQkXgChJ3NLXOYby9v8CUg/5bqr/Lto9K/ZyFgmYaAcOp1q2xHXW9rzx6/2ltmDqQuU1rIJRZ59GPyDj8CCqZbA8lw8lidrCMfEwH766kOQRrvmTuqzshhuh7NJY+AMCvapp2j3LvXANsn9GPyDk0BSRqUkXiZaOjO4WbkYNpcClBExN2Hj+PD84tX+aq6DqyoqwfB837EYWegbQPytNHB1hjfjaMpYJQMOFVSU0bi6uSLIBtBRkkBT9cWwqih9tvxLnZOI19prEmkmiJ9V01nlvgGv7xYLZERh6jzkiEMQvIO3UPQNTNDOmqypvpn6+ifW0LAlQGFWDDsPabES6nPq3fJPDpefkpneuW34bJaTaV2jWqIo4yy80ra54CGhgioSv1h3nPsq+TyEFQaeYvaNE3ScaIvJN1bLpVUTDL0lwoLbQHH9DXSJlkoGrCXOPhuB/EJPAsohROxdTpd0lhODTz3rpObfZTv/97zFpM5/FwF5c91BGRT+R1yXW+u8Cegl3lAdKBY41K14qZYc7Ly/lmL8XzFZAFf9QDlfapR6Psfvyn9e94E7DEn30DcLy5qj5FlXUEqcq0VYWYmc+IpJhsYDOSmIqhJp5kPmMflTUCzGtSkZovCFddE4y3EdxXNLHx8zLN1ZkMGfQGrfO27DOsmq75wZICE+TClvAmo0Z6axBZdmabJwuRmOTDJaYi1mF2bBl6gL2BCKHkh/iZ8OVR0acw4Lm8CmqsGmAsqemC3JsbSTjZhtWPYVxa2OnOhg74A71RqsqkqhEcdGcflTcCPYrKFxhaLolvPv9SeLBaZeTSPlFgaY+fF9AVUUPV1tdAdwkzm/UfzdxWUKA6MCTDvVbSsqDvR19IvTmJn5lT6iLdooS9gmH1GAVTssU9QHK3Wg3FcHusBlzrWitUYLSB/hoPYQgSAONwoOyugLyCvG5B4SUD7d9mgsZ4qaBkINoiPsk2rYVWtbWo/Msq+UpjVA/5aP3XNBeJP6zqLF80EE7CvbvPo314clbJ7i4Y/GFXE8m6zfrYkmIDhvRqQF5+TnY/rTYoEBgIyQogjatAednEFEzAoYilRId/UxDteoIA0oS9gD2iTenhzG7CfVVyhBLzvYGFRfVyV6Kp+0tZGeRagLyCMejdL2RXNi1k0UbZsKws66xDUx9K+sb3HeUFi0oPBOGLbqEk6u2qlQAL21lP8LbW0MJO0Dc75tZIn657WuIe+gKqqFv8Tq7OKK5CAkQsgvBNuZie2t7Zp6tpjjSBBaUFfwHSLxY8UjxabfxLD2Q4hBzK5VsFTelKRN8Zi9AhBgtKCvoD8QWZABMwT2b1iIJCAjdHkZ7ioNYQ5EV2qzhAkKC2Y1APuZa47wHZEEIEE5FabmAPffSGyHtte1vys2SX9OYTmU3o1kQkPu0sDpdHBdcWerYLMbISJSQtTGcpQNzlXc3McnFP6RMR3sXymP7nQmMpQhmXT0vWk8v16ebSOB9QoMfVDkIreCYFSacs/5hjhaxEm1XPux2T9QT2HHzUbvs2HMGGpIEFpYUL9hn7MzdayYLteTyE840P2lHJTfluAoDQxZQGvfBfnw3dJYUSNZbZrTOIgeQr/MWljygIWqx4MN9oHs6J9a3lb1ttqhJ2WmVDPuR/R+1tqMmU67DD4eD9ZV88QX+MbWd6Ur4L6f01NRs19bF+9mk3N6qvDfvDgYuwDTjFlAVsjyIeRuZW+DQGW7gPgHrkjjNnGf1h6mLKAgsiYc1nHGrZ27yj9UN91F4x1hWPm8x+WHqYsAL5fUEsWvqFhRoyVv38Hf1i/DuyaKkBYWpi0AIoCST+3SNfa5pKG3qN/dzW6ISXLgQBzB0+zZoMCQV2XTvK9QkU1GNMXcE0ccnOYq0OCxMy6yhzja6dn+gK6VnfYpzhWHwCJTUjTcKPrudL0BbhNb13X1t5eFJaiWO/Wjt2oBzzAjwBdDervz5+nxnIlnfLY4HjDY/1AazGwvgnhtNhGQoU1FH4E6GpQ/+/YQiTL6JTHhqa7zktlta1EvavnwtPBwUKFNRSeDkF6G9QLdwj6scKsRr9/ZeuQ3349PFa5q1BhDYWvc4C+BvUCPhE7IHOsKpNGxbYN2drZTnenBojgS4C+BvVCPpKcMvyf/MmONpYOlqKFwkU1ENO/CoLwcEgBXFCrjtiuwzSjOweXCwHKtu3+CtndvVEBVHrdEi6sYZQHATBvnj/wnkTWwRqfEDCsQZQLAQQ1z8B8srv728KG1U95ETDd21USsDKZedN+vignAt4EVK84fphMdlvQqIZQTgTM7wVPzxqzuYLxtZUvJwLaZ1CTfmsFjWoI5URAtKrvjq++FjSqIZQTATOoceU/VGI37AcflBMBz33m5MJ7XxjhoNrlRAD8r7ONp+NUI2yqXV4EvEyZMP+xsCENo5wIOODRfVScfIv+hIJTPgQ8ch3l4u0YJLsqZFDDKB8CVtVqdA3CDGk/IYMaRvkQMMKKGjqgj4+QQQ2jnAhQNekcYSdkUMMoHwL2m5NH/7tuHkIGNYzyISDP1nnImlHy9r30JxWa8iEArvP9X/zEJPebggY1iHIiAB5uJHXv8Z+wMQ2ivAgwWrAAxGABiMECEIMFIKbcCHgxN27wQYFjGkJ5EXDCPWHLipqd8vWnFJhyIuCD7yHiM78Fu37W+KCcCDgZRk2ORAga1RDKiYC9MdTkb92jfqKCPwHK1y/KGOBBYAGXA6ht2dlG0KiGwJOA62MrWwIg8R9zQ0cCgQUo680nDDwMZDnqAQ/wI+A3e88BSzelLhvk66DjVSjBX0upHza8l3yBsEENgR8BTSPUgwbmtdcxdovg9QDlT0tS7woc0xD4EWC/vnDuOwfN9bfGF7UTXk6nPBOGHwF14wrPvyO0Btq4XdhQfp4ruxFQTAd+BOwQRa07+ffVUxuixTtLT1F/3Fo9jInuQY+OrWlm6NKKZoYvW9DM0C1W369c68XPVVBmJKCIzNSRYN0AfYQ4VaOHpwPNDN62NDP42tDM4G+j92cONbQbQbr1gOeXDx++xKZ/qqXDaGZIpfvIfX87mhlO0G1idqUGzQxlwHFNWD9YgDZYABagByyAW7AAbbAALEAPWAC3YAHaCC5gVZL+NFps7Uszw/exNDOcbkozwz8hNDOUgeAC8rJpZvhAt2v6gpc0Myhf0MwAOewpWXABGG2wAMRgAYjBAhCDBSAGC0AMFoAYLAAxWABisADEYAGIwQIQI6SAnfUcmp0rXlQ0i6ORoWBRNZugFQWGFl8iFgfFf1yo/h9gAAIKyBR3X/eZw/Wi5a+Bnu3XyjBXNHRrgmi6galLxmJfPJMfYAgCCohspoRvPMYWLt6ytdWz/ZoZlLL+xOcQa92twbSKLxGLg+KZ/ABDEE7Ac/AN8TmgsnpREdmzUdnbr5XhPiBHy9sJdA4UoJW6RCwOimfyAwxCOAGXwUnic6ml+v3eVe4v9Gy/VobcqznE53CJzsGztVKXiMVB8Ux+gEEIJ+Aw+Jv43ARUD7huSfdCPduvnYFko5nuB8paqT/OyrZ4Jj/AIAQRoMjPzy84DMjurVIB9ThPEUlsu+7t/zgDwf1OoPt7nTG0UpfMqjeD3uJLZtDzAwxHEAFHAABJlwA51tQyCfUfnCK78fJlvc4vdYzC+XEGCHc4+meUEUMrdYms+jPoLb5kBj0/wHAEEfDm4sWL95+DjcRsoj+1Jkn1mjvQ0bru4wzECbJvmT0Ua6XWzmpABr3Fl8yg5wcYjoCXoU2jIcyrOJqav3GUIKj50WeGZngvjy/rz7lEau0FTopn8gMMQciKmGhk5hcONyBcGKV6NUXfIVQzw08gIZlE97i1WsUXLXBWPJMfYAhC3orYHmYfeZ6YxgPVmzt6t18jwzfqf/lHBhZfuMBd8Ux+gAHgm3GIwQIQgwUgBgtADBaAGCwAMVgAYrAAxGABiMECEIMFIAYLQAwWgBgsADFYAGKwAMRgAYjBAhCDBSAGC0AMFoAYLAAxWABisADEYAGIwQIQg0pAM1B2Ay4d7E6H0De5xEq3FdopskHvMgtZ7axK8k9NyX9PG1meYLIhnIFIwEOxrZ/+15E/Jq69PgFEipyo+WUWQgigkox0y8xdbrnrFYPt4A5EApaarwZnGeQzSIA+CAEUX0VBOLcSg63gEkQC6sW8sxlBTGNrER8Kz/7w/Ugfn4mrNPrjrLRxuKf/6scxjr7bIVSur2MTTBx96gEAXvoumF7JvuMLOMaT+B967zCHEvB6oKfEb5pCnYJQVDCzmm1dsumj7+bkavad1R0jvunv5ZW4zJmy2Er9RvTRouKh79bJTteLlwpz5k+uYtvkbPFmcAkaATfBDtjVSwFhGrgD4SlwDHazX5ga6qspwHvYofbAZ+YPdaV5cJFocHoPsAXejYm6XuDrGfvdNIvB8FdwAcLvwXVKQLxsSloCSFOnIAQkmk/c1g1kELuxUcyBFZJBVKHKpjYzNzZwUAl48GWD69eTvK/nFBUPfZtVGvqieKkwZ7ztgm2fyR4Vp+MQNAJm2+fA3YA4+70mR/1J8lb8BXZD+MpeU0ArCK+AYcRZFVzLdiDP2KsYE20AAAMvSURBVF0rqw9B1QlzcQ1hget0CPvVUR2CviCHbwoeXXSQume+mNjfbUKIhcACCHurxrw5CnZBmOutElB0CCou3rfiO80ldc7r4q3ExlmnFH/DIWgEBHXPzn5qk0jMtYuCyorj4HoJ2U1DZ00BxN7NB8Th509w9Sy4QqzZJ/6g2r1kI6GkehD2qUtImFt0DniYZplUJOAA1dYiVZSnSj42lEqSbKcg85YQUFy870itJXXOLSKy9eT9rOJvOASJgD9VB183Yqenmr/8HfwJp3qR64dpCphLCthDCdilPlrfUe1ectw2UsBe8PAXcEMl4HIbd9e2bsUC1onIXX2YyEIlVwtIqkp+Li4hoLh433laS+qcc9Un7eJvOASJgAlWR0+cODEP/ARhlkXaROI4scaS3F/ddAj4FRy6SPKu+CqIFPDWav0QcscSAl7b9L2khHWLBWRS/wGbQY4quVrAYuo/YHQJAcXFk2u1l8icG83ILiRu3S/+hkNQCFD6dyAnbyRkBxmtuwQSf2gXyL4asp10CMiy3kCsWdNHqS0AtoupQPzJkgIOg7sQPpMWC7hntoQ8BwRBLQEnyHNAnm8JAcXFk2u1l8icf4CdxJnDdX7xNxyCQsAZsJWaRjsTR9MUiegeMd/eYdHmetWDilsgagqAk22m7hgtnklcj1T56UOxgHUicBNSAm6ZxWxaVsUx9FJRikEWk9VXQRoClFE2M1Mb+5UQUFw8lVZ7iczZxS55Zxv7/4q/4RAUAoZJVN0tbADfQ/jcLJKcf/eVW5VlC5sUt0DUEqBcGmQduJT4y/u1ku2rYgGPRdSAzuQ5IK2ytOHBQx7JRSnypwdKVfUADQEwu38Fj4QDJQUUFU+l1V4ic+Yl+dtGnNX4hkOM5GZcfjZ5cB7ZiWYuh7LvOXwKGImA6+AXokYcQPO/+0CZ/ft8GhiJAGXj6gd/6Sp5QCfPi2VVO/C1PcJhJALgozhXh0ZnaGW569zsHk9bIyDGIqDcggUgBgtADBaAGCwAMVgAYrAAxGABiMECEIMFIAYLQAwWgBgsADFYAGKwAMRgAYjBAhCDBSAGC0AMFoAYLAAx/we0qJv9NVVgewAAAABJRU5ErkJggg==" style="display: block; margin: auto;" /></p>
<p>The <code>tcga_colon_curves</code> object is another <code>GRanges</code> object which contains the basepair resolution methylation data used to define these regions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">show</span>(tcga_colon_curves)</code></pre></div>
<pre><code>## GRanges object with 7135 ranges and 7 metadata columns:
##          seqnames                 ranges strand |        id    type
##             &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;array&gt;
##      [1]    chr11       [131996, 132411]      * |    129466 OpenSea
##      [2]    chr11       [189654, 189654]      * |    129467 OpenSea
##      [3]    chr11       [190242, 190242]      * |    129468 OpenSea
##      [4]    chr11       [192096, 192141]      * |    129469 OpenSea
##      [5]    chr11       [192763, 193112]      * |    129470 OpenSea
##      ...      ...                    ...    ... .       ...     ...
##   [7131]    chr11 [134892703, 134892703]      * |    142360 OpenSea
##   [7132]    chr11 [134903175, 134903175]      * |    142361 OpenSea
##   [7133]    chr11 [134910774, 134910774]      * |    142362 OpenSea
##   [7134]    chr11 [134911302, 134911302]      * |    142363 OpenSea
##   [7135]    chr11 [134945848, 134945848]      * |    142364 OpenSea
##          blockgroup         diff      smooth normalMean cancerMean
##           &lt;numeric&gt;     &lt;matrix&gt;   &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;
##      [1]        495 -0.088871013 -0.12366708 0.75449261 0.66562159
##      [2]        495 -0.001244860 -0.08656241 0.95952884 0.95828398
##      [3]        495 -0.164759184 -0.08625123 0.68006830 0.51530911
##      [4]        495  0.003094723 -0.08526643 0.05134196 0.05443669
##      [5]        495 -0.102463563 -0.08484055 0.55235839 0.44989482
##      ...        ...          ...         ...        ...        ...
##   [7131]        520  -0.11394759  -0.1421404  0.4013819  0.2874343
##   [7132]        520  -0.01148901  -0.1530226  0.8482137  0.8367247
##   [7133]        520  -0.20066455  -0.1627672  0.6365234  0.4358588
##   [7134]        520  -0.14013791  -0.1635050  0.5831618  0.4430239
##   [7135]        520  -0.24670370  -0.2308602  0.6757470  0.4290433
##   -------
##   seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
<p>This basepair resolution data includes mean methylation levels for normal and cancer and a smoothed estimate of methylation difference. This smoothed difference estimate is used to define regions in the <code>tcga_colon_blocks</code> object.</p>
<p>Finally, the <code>apColonData</code> object is an <code>ExpressionSet</code> containing gene expression data for colon normal and tumor samples for genes within regions of methylation loss identified <a href="http://www.nature.com/ng/journal/v43/n8/full/ng.865.html">this paper</a>. Our goal in this vignette is to visualize this data as we browse the genome.</p>
</div>
<div id="the-epivizr-app" class="section level3">
<h3>The epivizr app</h3>
<p>The connection to <code>Epiviz</code> is managed through a session manager object of class <code>EpivizApp</code>. We can create this object and open <code>Epiviz</code> using the <code>startEpiviz</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app &lt;-<span class="st"> </span><span class="kw">startEpiviz</span>(<span class="dt">workspace=</span><span class="st">&quot;qyOTB6vVnff&quot;</span>, <span class="dt">gists=</span><span class="st">&quot;2caf8e891201130c7daa&quot;</span>)</code></pre></div>
<pre><code>## .</code></pre>
<p>This opens a websocket connection between the interactive <code>R</code> session and the browser client. This will allow us to visualize data stored in the <code>Epiviz</code> server along with data in the interactive <code>R</code> session.</p>
<hr />
<p><em>Windows users:</em> In Windows platforms we need to use the <code>service</code> function to let the interactive <code>R</code> session connect to the <code>epiviz</code> web app and serve data requests. We then escape (using <code>ctl-c</code> or <code>esc</code> depending on your environment) to continue with the interactive <code>R</code> session. This is required anytime you want <code>epivizr</code> to serve data to the web app, for example, when interacting with the UI. (We are actively developing support for non-blocking sessions in Windows platforms).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$server$<span class="kw">service</span>()</code></pre></div>
<hr />
</div>
<div id="listing-available-chart-types" class="section level3">
<h3>Listing available chart types</h3>
<p>Once the browser is open and is connected to an active epivizr session, the epivizr session registers available chart types supported in the epiviz JS app. To list available chart types we use method <code>list_chart_types</code>. In this vignette, we only show three chart types available for use in the epiviz JS app. The production epiviz JS app has a larger number of chart types available. Chart types can also be added dynamically as described in the epiviz JS documentation: <a href="http://epiviz.cbcb.umd.edu/help" class="uri">http://epiviz.cbcb.umd.edu/help</a>. Chart types that are added dynamically are listed and usable within an <code>epivizr</code> session.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$chart_mgr$<span class="kw">list_chart_types</span>()</code></pre></div>
<pre><code>##          type                          js_class num_settings
## 1 BlocksTrack epiviz.plugins.charts.BlocksTrack            6
## 2   LineTrack   epiviz.plugins.charts.LineTrack           15
## 3 ScatterPlot epiviz.plugins.charts.ScatterPlot           11
##                                                                              settings
## 1                title,marginTop,marginBottom,marginLeft,marginRight,minBlockDistance
## 2 title,marginTop,marginBottom,marginLeft,marginRight,measurementGroupsAggregator,...
## 3 title,marginTop,marginBottom,marginLeft,marginRight,measurementGroupsAggregator,...
##   num_colors
## 1         10
## 2          8
## 3         10</code></pre>
<p>Chart type settings, e.g., line widths, colors, margins etc. can be modified dynamically. Settings available for each chart type are briefly listed in this call. See below for further detail on customizing charts through settings and colors.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$chart_mgr$<span class="kw">print_chart_type_info</span>(<span class="st">'BlocksTrack'</span>)</code></pre></div>
<pre><code>## Settings for chart type  BlocksTrack 
##                 id                  label default_value possible_values
## 1            title                  Title                              
## 2        marginTop             Top margin            25                
## 3     marginBottom          Bottom margin            23                
## 4       marginLeft            Left margin            20                
## 5      marginRight           Right margin            10                
## 6 minBlockDistance Minimum block distance             5                
##     type
## 1 string
## 2 number
## 3 number
## 4 number
## 5 number
## 6 number
## Colors: #1f77b4, #ff7f0e, #2ca02c, #d62728, #9467bd, #8c564b, #e377c2, #7f7f7f, #bcbd22, #17becf</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$chart_mgr$<span class="kw">list_chart_type_settings</span>(<span class="st">'LinePlot'</span>)</code></pre></div>
<pre><code>## data frame with 0 columns and 0 rows</code></pre>
</div>
<div id="adding-block-region-devices" class="section level3">
<h3>Adding block region devices</h3>
<p>Once the browser is open we can visualize the <code>tcga_colon_blocks</code> object containing blocks of methylation modifications in colon cancer. We use the <code>plot</code> method to do so.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blocks_chart &lt;-<span class="st"> </span>app$<span class="kw">plot</span>(tcga_colon_blocks, <span class="dt">datasource_name=</span><span class="st">&quot;450k colon_blocks&quot;</span>)</code></pre></div>
<hr />
<p><em>Windows users:</em> We need the <code>service</code> function to let the interactive <code>R</code> session serve data requests from the browser client as you interact with the UI. Escape (using <code>ctl-c</code> or <code>esc</code> depending on your environment) to continue with the interactive <code>R</code> session.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$server$<span class="kw">service</span>()</code></pre></div>
<hr />
<p>You should now see that a new track is added to the <code>Epiviz</code> web app. You can think of this track as an interactive display device in <code>R</code>. As you navigate on the browser, data is requested from the <code>R</code> session through the websocket connection. Remember to escape to continue with your interactive <code>R</code> session if you are not running “non-blocking” mode. The <code>blocks_chart</code> object inherits from class <code>EpivizChart</code>, which we can use to get information about the data being displayed and the chart used to display it. Note that the “brushing” effect we implement in <code>Epiviz</code> works for <code>epivizr</code> tracks as well.</p>
<p>Now that we have interactive data connections to <code>Epiviz</code> we may want to iteratively compute and visualize our data. For example, we may want to only display methylation blocks inferred at a certain statistical significance level. In this case, we will filter by block size.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset to those with length &gt; 250Kbp</span>
keep &lt;-<span class="st"> </span><span class="kw">width</span>(tcga_colon_blocks) &gt;<span class="st"> </span><span class="dv">250000</span>

<span class="co"># get the data object for chart</span>
ms_obj &lt;-<span class="st"> </span>app$<span class="kw">get_ms_object</span>(blocks_chart)
app$<span class="kw">update_measurements</span>(ms_obj, tcga_colon_blocks[keep,])</code></pre></div>
<p>Now, only this subset of blocks will be displayed in the already existing track.</p>
</div>
<div id="modifying-chart-settings-and-colors" class="section level3">
<h3>Modifying chart settings and colors</h3>
<p>To modify default chart colors or settings, we need to know what settings can be applied to the chart. Again, as these are defined dynamically in the epiviz JS app we use a method, <code>print_chart_type_info</code> on the <code>EpivizChartMgr</code> class to list settings that can be applied to specific chart type. For example, to list chart settings available for a <code>BlocksTrack</code> chart we use the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$chart_mgr$<span class="kw">print_chart_type_info</span>(<span class="st">&quot;BlocksTrack&quot;</span>)</code></pre></div>
<pre><code>## Settings for chart type  BlocksTrack 
##                 id                  label default_value possible_values
## 1            title                  Title                              
## 2        marginTop             Top margin            25                
## 3     marginBottom          Bottom margin            23                
## 4       marginLeft            Left margin            20                
## 5      marginRight           Right margin            10                
## 6 minBlockDistance Minimum block distance             5                
##     type
## 1 string
## 2 number
## 3 number
## 4 number
## 5 number
## 6 number
## Colors: #1f77b4, #ff7f0e, #2ca02c, #d62728, #9467bd, #8c564b, #e377c2, #7f7f7f, #bcbd22, #17becf</code></pre>
<p>There are two ways to set settings and colors to a chart: when the chart is initially created using the <code>plot</code> method, or after the chart is created using the <code>set</code> method in class <code>EpivizChart</code>. We will illustrate both ways of doing this:</p>
<p>If using the <code>plot</code> method, use the <code>list_chart_settings</code> method as above to list settings that can be applied to a chart type. For example, for a <code>BlocksTrack</code> chart, we can set <code>minBlockDistance</code> as a setting for the plot function which controls how close genomic regions can be before they are merged into a single rectangle in the chart. We can also change colors used in the chart this way.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">settings &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">minBlockDistance=</span><span class="dv">50</span>)
colors &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#d15014&quot;</span>, <span class="st">&quot;#5e97eb&quot;</span>, <span class="st">&quot;#e81ccd&quot;</span>)
blocks_chart_wset &lt;-<span class="st"> </span>app$<span class="kw">plot</span>(tcga_colon_blocks, <span class="dt">datasource_name=</span><span class="st">&quot;450k colon_blocks&quot;</span>, <span class="dt">settings=</span>settings, <span class="dt">colors=</span>colors)</code></pre></div>
<p>This will create a second blocks track using a different color map.</p>
<p>On the other hand, if the <code>BlockChart</code> is already added to the epiviz JS application session, use the <code>set_chart_settings</code> method to update settings and colors.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create a list of settings and colors to update the plot</span>
settings &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">minBlockDistance=</span><span class="dv">100</span>)
colors &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#5e97eb&quot;</span>, <span class="st">&quot;#d15014&quot;</span>, <span class="st">&quot;#e81ccd&quot;</span>)
app$chart_mgr$<span class="kw">set_chart_settings</span>(blocks_chart_wset, <span class="dt">settings=</span>settings, <span class="dt">colors=</span>colors)</code></pre></div>
<p>This changes the color map for the second blocks track added. Use the <code>print_chart_info</code> method to list settings and colors currently used in the chart.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># to list applied chart settings</span>
app$chart_mgr$<span class="kw">print_chart_info</span>(blocks_chart_wset)</code></pre></div>
<pre><code>## Chart settings for chart id  epivizChart_2 :
##                 id                  label default_value possible_values
## 1            title                  Title                              
## 2        marginTop             Top margin            25                
## 3     marginBottom          Bottom margin            23                
## 4       marginLeft            Left margin            20                
## 5      marginRight           Right margin            10                
## 6 minBlockDistance Minimum block distance           100                
##     type
## 1 string
## 2 number
## 3 number
## 4 number
## 5 number
## 6 number
## Colors: #5e97eb, #d15014, #e81ccd</code></pre>
<p>Methods are also available directly on the <code>EpivizChart</code> objects. E.g., calls <code>blocks_chart$set(settings=settings, colors=colors)</code> and <code>blocks_chart$print_info()</code> yield the same result.</p>
</div>
<div id="printing-charts" class="section level3">
<h3>Printing charts</h3>
<p>Charts can be printed as a pdf or png file through the epivizr session. To do so, use the <code>print_chart</code> method:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$chart_mgr$<span class="kw">print_chart</span>(blocks_chart_wset, <span class="dt">file_name=</span><span class="st">&quot;blocks_chart&quot;</span>, <span class="dt">file_type=</span><span class="st">&quot;pdf&quot;</span>)</code></pre></div>
<p>This will create a file named <code>blocks_chart.pdf</code> which will be downloaded through your web browser. It will be found in the default file download location for your web browser.</p>
</div>
<div id="adding-line-plots-along-the-genome" class="section level3">
<h3>Adding line plots along the genome</h3>
<p>There are a number of different data types available to use through <code>epivizr</code>. You can see a list of R/BioC classes that are supported using <code>?register</code>. In the previous section, we used the <code>block</code> data type to register a <code>GenomicRanges</code> object. To visualize methylation data at base-pair resolution from the <code>tcga_colon_curves</code> <code>GenomicRanges</code> object, we will use the <code>bp</code> type.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># add low-filter smoothed methylation estimates</span>
means_track &lt;-<span class="st"> </span>app$<span class="kw">plot</span>(tcga_colon_curves, <span class="dt">datasource_name=</span><span class="st">&quot;450kMeth&quot;</span>,<span class="dt">type=</span><span class="st">&quot;bp&quot;</span>,<span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;cancerMean&quot;</span>,<span class="st">&quot;normalMean&quot;</span>))</code></pre></div>
<p>NOTE: You can adjust track settings to change how this new track looks like. For instance, to show all points in the window set the <code>step</code> parameter to 1, and to see a smooth interpolation of the data set the <code>interpolation</code> parameter to <code>basis</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">means_track$<span class="kw">set</span>(<span class="dt">settings=</span><span class="kw">list</span>(<span class="dt">step=</span><span class="dv">1</span>, <span class="dt">interpolation=</span><span class="st">&quot;basis&quot;</span>))</code></pre></div>
<p>Notice that we added two lines in this plot, one for mean methylation in cancer and another for mean methylation in normal. The <code>columns</code> argument specifies which columns in <code>mcols(colon_curves)</code> will be displayed.</p>
<p>We can also add a track containing the smooth methylation difference estimate used to define methylation blocks.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">diff_chart &lt;-<span class="st"> </span>app$<span class="kw">plot</span>(tcga_colon_curves, <span class="dt">datasource_name=</span><span class="st">&quot;450kMethDiff&quot;</span>,<span class="dt">type=</span><span class="st">&quot;bp&quot;</span>,<span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;smooth&quot;</span>),<span class="dt">ylim=</span><span class="kw">matrix</span>(<span class="kw">c</span>(-.<span class="dv">5</span>,.<span class="dv">5</span>),<span class="dt">nc=</span><span class="dv">1</span>))</code></pre></div>
<p>We pass limits for the y axis in this case. To see other arguments supported, you can use the help framework in R <code>?&quot;EpivizApp&quot;</code>. As before, we can specify settings for this new track.</p>
</div>
<div id="the-app-connection-again" class="section level3">
<h3>The app connection again</h3>
<p>We can use the app connection object to list charts we have added so far, or to remove charts.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$chart_mgr$<span class="kw">list_charts</span>()</code></pre></div>
<pre><code>##              id                              type
## 1 epivizChart_1 epiviz.plugins.charts.BlocksTrack
## 2 epivizChart_2 epiviz.plugins.charts.BlocksTrack
## 3 epivizChart_3   epiviz.plugins.charts.LineTrack
## 4 epivizChart_4   epiviz.plugins.charts.LineTrack
##                                  measurements connected
## 1       450k colon_blocks_1:450k colon_blocks          
## 2       450k colon_blocks_2:450k colon_blocks          
## 3 450kMeth_3:cancerMean,450kMeth_3:normalMean          
## 4                       450kMethDiff_4:smooth</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$chart_mgr$<span class="kw">rm_chart</span>(means_track)
app$chart_mgr$<span class="kw">list_charts</span>()</code></pre></div>
<pre><code>##              id                              type
## 1 epivizChart_1 epiviz.plugins.charts.BlocksTrack
## 2 epivizChart_2 epiviz.plugins.charts.BlocksTrack
## 3 epivizChart_4   epiviz.plugins.charts.LineTrack
##                            measurements connected
## 1 450k colon_blocks_1:450k colon_blocks          
## 2 450k colon_blocks_2:450k colon_blocks          
## 3                 450kMethDiff_4:smooth</code></pre>
</div>
<div id="adding-a-scatterplot" class="section level3">
<h3>Adding a scatterplot</h3>
<p>Now we want to visualize the colon expression data in <code>apColonData</code> object as an MA plot in <code>Epiviz</code>. First, we add an <code>&quot;MA&quot;</code> assay to the <code>ExpressionSet</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">keep &lt;-<span class="st"> </span><span class="kw">pData</span>(apColonData)$SubType!=<span class="st">&quot;adenoma&quot;</span>
apColonData &lt;-<span class="st"> </span>apColonData[,keep]
status &lt;-<span class="st"> </span><span class="kw">pData</span>(apColonData)$Status
Indexes &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">seq</span>(<span class="dt">along=</span>status),status)

exprMat &lt;-<span class="st"> </span><span class="kw">exprs</span>(apColonData)
mns &lt;-<span class="st"> </span><span class="kw">sapply</span>(Indexes, function(ind) <span class="kw">rowMeans</span>(exprMat[,ind]))
mat &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">colonM=</span>mns[,<span class="st">&quot;1&quot;</span>]-mns[,<span class="st">&quot;0&quot;</span>], <span class="dt">colonA=</span><span class="fl">0.5</span>*(mns[,<span class="st">&quot;1&quot;</span>]+mns[,<span class="st">&quot;0&quot;</span>]))
<span class="kw">assayDataElement</span>(apColonData, <span class="st">&quot;MA&quot;</span>) &lt;-<span class="st"> </span>mat
<span class="kw">show</span>(apColonData)</code></pre></div>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 5339 features, 53 samples 
##   element names: exprs, MA 
## protocolData: none
## phenoData
##   sampleNames: GSM95473 GSM95474 ... GSM215114 (53 total)
##   varLabels: filename DB_ID ... Status (7 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use 'experimentData(object)'
## Annotation: hgu133plus2</code></pre>
<p><code>epivizr</code> will use the <code>annotation(apColonData)</code> annotation to determine genomic locations using the <code>AnnotationDbi</code> package so that only probesets inside the current browser window are displayed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">eset_chart &lt;-<span class="st"> </span>app$<span class="kw">plot</span>(apColonData, <span class="dt">datasource_name=</span><span class="st">&quot;MAPlot&quot;</span>, <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;colonA&quot;</span>,<span class="st">&quot;colonM&quot;</span>), <span class="dt">assay=</span><span class="st">&quot;MA&quot;</span>)</code></pre></div>
<pre><code>## Loading required package: hgu133plus2.db</code></pre>
<pre><code>## Loading required package: AnnotationDbi</code></pre>
<pre><code>## Loading required package: org.Hs.eg.db</code></pre>
<pre><code>## </code></pre>
<pre><code>## </code></pre>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<p>In this case, we specify which data is displayed in each axis of the scatter plot using the <code>columns</code> argument. The <code>assay</code> arguments indicates where data is obtained.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$chart_mgr$<span class="kw">rm_chart</span>(eset_chart)</code></pre></div>
</div>
<div id="the-rangedsummarizedexperiment-object" class="section level3">
<h3>The RangedSummarizedExperiment Object</h3>
<p><code>Epiviz</code> is also able to display plots of data in the form of a <code>RangedSummarizedExperiment</code> object. After loading the <code>tcga_colon_expression</code> dataset in the <code>epivizrData</code> package, we can see that this object contains information on 239322 exons in 40 samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(tcga_colon_expression)
<span class="kw">show</span>(tcga_colon_expression)</code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 12800 40 
## metadata(0):
## assays(1): ''
## rownames(12800): 143686 149186 ... 149184 149185
## rowData names(2): exon_id gene_id
## colnames(40): TCGA-A6-5659-01A-01R-1653-07
##   TCGA-A6-5659-11A-01R-1653-07 ... TCGA-D5-5540-01A-01R-1653-07
##   TCGA-D5-5541-01A-01R-1653-07
## colData names(110): bcr_patient_barcode bcr_sample_uuid ...
##   Basename fullID</code></pre>
<p>The <code>assay</code> slot holds a matrix of raw sequencing counts, so before we can plot a scatterplot showing expression, we must first normalize the count data. We use the geometric mean of each row as a reference sample to divide each column (sample) by, then use the median of each column as a scaling factor to divide each row (exon) by.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ref_sample &lt;-<span class="st"> </span><span class="dv">2</span> ^<span class="st"> </span><span class="kw">rowMeans</span>(<span class="kw">log2</span>(<span class="kw">assay</span>(tcga_colon_expression) +<span class="st"> </span><span class="dv">1</span>))
scaled &lt;-<span class="st"> </span>(<span class="kw">assay</span>(tcga_colon_expression) +<span class="st"> </span><span class="dv">1</span>) /<span class="st"> </span>ref_sample
scaleFactor &lt;-<span class="st"> </span>Biobase::<span class="kw">rowMedians</span>(<span class="kw">t</span>(scaled))
assay_normalized &lt;-<span class="st"> </span><span class="kw">sweep</span>(<span class="kw">assay</span>(tcga_colon_expression), <span class="dv">2</span>, scaleFactor, <span class="st">&quot;/&quot;</span>)
<span class="kw">assay</span>(tcga_colon_expression) &lt;-<span class="st"> </span>assay_normalized</code></pre></div>
<p>Now, using the expression data in the <code>assay</code> slot and the sample data in the <code>colData</code> slot, we can compute mean exon expression by sample type.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">status &lt;-<span class="st"> </span><span class="kw">colData</span>(tcga_colon_expression)$sample_type
index &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">seq</span>(<span class="dt">along =</span> status), status)
logCounts &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">assay</span>(tcga_colon_expression) +<span class="st"> </span><span class="dv">1</span>)
means &lt;-<span class="st"> </span><span class="kw">sapply</span>(index, function(ind) <span class="kw">rowMeans</span>(logCounts[, ind]))
mat &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">cancer =</span> means[, <span class="st">&quot;Primary Tumor&quot;</span>], <span class="dt">normal =</span> means[, <span class="st">&quot;Solid Tissue Normal&quot;</span>])</code></pre></div>
<p>Now, create a new <code>RangedSummarizedExperiment</code> object with the two column matrix, and all the information about the features of interest, in this case exons, are stored in the <code>rowRanges</code> slot to be queried by <code>Epiviz</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sumexp &lt;-<span class="st"> </span><span class="kw">SummarizedExperiment</span>(mat, <span class="dt">rowRanges=</span><span class="kw">rowRanges</span>(tcga_colon_expression))
se_chart &lt;-<span class="st"> </span>app$<span class="kw">plot</span>(sumexp, <span class="dt">datasource_name=</span><span class="st">&quot;Mean by Sample Type&quot;</span>, <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;normal&quot;</span>, <span class="st">&quot;cancer&quot;</span>))</code></pre></div>
<p>Again, the <code>columns</code> argument specifies what data will be displayed along which axis.</p>
</div>
<div id="slideshow" class="section level3">
<h3>Slideshow</h3>
<p>We can navigate to a location on the genome using the <code>navigate</code> method of the app object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$<span class="kw">navigate</span>(<span class="st">&quot;chr11&quot;</span>, <span class="dv">110000000</span>, <span class="dv">120000000</span>)</code></pre></div>
<p>There is a convenience function to quickly navigate to a series of locations in succession. We can use that to browse the genome along a ranked list of regions. Let’s navigate to the 5 most up-regulated exons in the colon exon expression data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">foldChange &lt;-<span class="st"> </span>mat[,<span class="st">&quot;cancer&quot;</span>]-mat[,<span class="st">&quot;normal&quot;</span>]
ind &lt;-<span class="st"> </span><span class="kw">order</span>(foldChange,<span class="dt">decreasing=</span><span class="ot">TRUE</span>)

<span class="co"># bounding 1Mb around each exon</span>
slideshowRegions &lt;-<span class="st"> </span><span class="kw">trim</span>(<span class="kw">rowRanges</span>(sumexp)[ind] +<span class="st"> </span>1000000L)
app$<span class="kw">slideshow</span>(slideshowRegions, <span class="dt">n=</span><span class="dv">5</span>)</code></pre></div>
</div>
<div id="printing-the-epivizr-workspace" class="section level3">
<h3>Printing the epivizr workspace</h3>
<p>To print the current epiviz workspace as a pdf or png, we use the <code>print_workspace</code> method:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$<span class="kw">print_workspace</span>(<span class="dt">file_name=</span><span class="st">&quot;workspace&quot;</span>, <span class="dt">file_type=</span><span class="st">&quot;pdf&quot;</span>)</code></pre></div>
<p>This will create a file named <code>workspace.pdf</code> and will be downloaded through your web browser. It will be saved to the default file download location for your web browser.</p>
</div>
<div id="closing-the-session" class="section level3">
<h3>Closing the session</h3>
<p>To close the connection to <code>Epiviz</code> and remove all tracks added during the interactive session, we use the <code>stop_app</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$<span class="kw">stop_app</span>()</code></pre></div>
</div>
<div id="standalone-version-and-browsing-arbitrary-genomes" class="section level3">
<h3>Standalone version and browsing arbitrary genomes</h3>
<p>The <code>epivizrStandalone</code> package makes it possible to run the web app UI locally completely within R. The <code>epivizr</code> package implements two-way communication between the <code>R/Bioconductor</code> computational genomics environment and <code>Epiviz</code>. Objects in an <code>R</code> session can be displayed as tracks or plots on Epiviz.</p>
<p>To install the <code>epivizrStandalone</code> package, use the <code>biocLite</code> function again.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biocLite</span>(<span class="st">&quot;epivizrStandalone&quot;</span>)</code></pre></div>
<p>Running the epiviz visualization app as a standalone allows it to browse any genome of interest using Bioconductor infrastructure. For example, to browse the yeast genome we can install the Bioconductor package containing gene information for yeast (this might take a minute):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biocLite</span>(<span class="st">&quot;TxDb.Scerevisiae.UCSC.sacCer3.sgdGene&quot;</span>)</code></pre></div>
<p>Once this package is installed we can start a genome browser for yeast based on this annotation with a couple of lines of code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(epivizrStandalone)
<span class="kw">library</span>(TxDb.Scerevisiae.UCSC.sacCer3.sgdGene)

txdb &lt;-<span class="st"> </span>TxDb.Scerevisiae.UCSC.sacCer3.sgdGene
app &lt;-<span class="st"> </span><span class="kw">startStandalone</span>(txdb)    </code></pre></div>
<p>Make sure to stop the app when you are done.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$<span class="kw">stop_app</span>()</code></pre></div>
</div>
<div id="sessioninfo" class="section level3">
<h3>SessionInfo</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.3.0 (2016-05-03)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] hgu133plus2.db_3.2.3       org.Hs.eg.db_3.3.0        
##  [3] AnnotationDbi_1.34.3       SummarizedExperiment_1.2.3
##  [5] GenomicRanges_1.24.1       GenomeInfoDb_1.8.1        
##  [7] IRanges_2.6.0              S4Vectors_0.10.1          
##  [9] antiProfilesData_1.8.0     Biobase_2.32.0            
## [11] BiocGenerics_0.18.0        epivizr_2.2.3             
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.5             BiocInstaller_1.22.2   
##  [3] formatR_1.4             XVector_0.12.0         
##  [5] GenomicFeatures_1.24.2  bitops_1.0-6           
##  [7] tools_3.3.0             zlibbioc_1.18.0        
##  [9] biomaRt_2.28.0          digest_0.6.9           
## [11] RSQLite_1.0.0           evaluate_0.9           
## [13] graph_1.50.0            DBI_0.4-1              
## [15] OrganismDbi_1.14.1      yaml_2.1.13            
## [17] epivizrData_1.0.3       rtracklayer_1.32.1     
## [19] stringr_1.0.0           knitr_1.13             
## [21] Biostrings_2.40.2       R6_2.1.2               
## [23] XML_3.98-1.4            RBGL_1.48.1            
## [25] BiocParallel_1.6.2      rmarkdown_0.9.6        
## [27] magrittr_1.5            epivizrServer_1.0.3    
## [29] Rsamtools_1.24.0        htmltools_0.3.5        
## [31] GenomicAlignments_1.8.1 mime_0.4               
## [33] httpuv_1.3.3            stringi_1.1.1          
## [35] RCurl_1.95-4.8          rjson_0.2.15</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
